veps/sig-compute/target-xml-mutation-hooks.md

```diff
@@ -0,0 +1,141 @@
+
+## Goals
+
+- Streamline the overall migration process.
+- Mitigate the growth of the VMI.Status field.
+- Enable migration capabilities for new features.
+- Enable upgrade paths for existing features.
+
+## Non Goals
+
+- Rewriting migration architecture
+- Introducing new API fields
+- Changing user-visible migration behavior
+
+## Definition of Users
+
+- KubeVirt maintainers
+- SIG Compute, Storage, Network
+- Contributors implementing migration-related features
+
+## User Stories
+
+- As a KubeVirt contributor, I want all XML mutations to happen on the target virt-launcher pod via hooks, so 
+migration-related features can be implemented safely.
+- As a KubeVirt user, I want live migrations to work reliably across upgrades, so VMs maintain correct state without 
+extra steps.
+- As a KubeVirt maintainer, I want to avoid exposing internal implementation details through vmi.status, keeping the API 
+clean and maintainable.
+
+## Repos
+
+- `kubevirt/kubevirt`
+
+## Design
+
+Libvirt hooks allow scripts to run at key points in a VM’s lifecycle, such as start, stop, or migration. 
+KubeVirt uses them [to mark](https://github.com/kubevirt/kubevirt/blob/release-1.7/cmd/virt-launcher/qemu) 
+which backend storage PVC contains the correct persistent state. 
+As part of this design, we want to explore using these hooks to move pre-migration XML 
+modifications to the target virt-launcher pod.
+For more details, see [Libvirt Hooks](https://libvirt.org/hooks.html#etc-libvirt-hooks-qemu)
+
+1. Target virt-launcher starts a lightweight hook server on a Unix domain socket.
+2. Libvirt executes a hook script during `migrate-begin` and passes domain XML to stdin.
+3. The script sends XML to the virt-launcher socket via `ncat`.
+4. Launcher applies all registered hooks (per SIG).
+5. Modified XML is returned to the hook script.
+6. Script writes final XML to stdout.
+7. Libvirt defines the domain using the modified XML.
+
+### Deprecation Plan
+
+Fields that become unnecessary:
+
+- `vmi.Status.MigrationState.TargetCPUSet`
+- `vmi.Status.MigrationState.TargetNodeTopology`
+
+Plan:
+1. Stop using these fields when modifying the xml and mark them as deprecated.
+2. Continue populating for one release for compatibility with older launcher images.
+3. Remove completely in the following release.
+
+### Alternatives
+
+#### Target-Initiated Migration
+
+Avoids hooks but requires:
+
+- Complete rewrite of migration architecture  
+- Supporting both source- and target-initiated paths  
+- Reworking live migration monitoring from source → target  
+
+Too complex.
+
+## Scalability
+
+It should be a scalable hook system that allows each SIG to maintain its own hooks.
+
+## Update / Compatibility
+
+- No API changes  
+- No feature gate   
+- should be safe during upgrade since the final XML should be the same as before
+
+## Functional Testing
+
+- Existing functional tests already cover XML mutation behavior
```

Comment:
This might be slightly out of scope for this VEP, but we need to consider the desire to move away from XML inspection in e2e tests. To achieve that, we probably need to keep/introduce VMI status fields that test code could look at instead of the libvirt domxml.

Reply:
Thanks for raising this. You’re right that finding an API to replace XML inspection in e2e tests will be challenging. It’s important, but probably belongs in a future KEP where we can design it properly.